[docs] Add Grid version 2 docs (#33554)
mui/material-ui: docs/data/material/components/grid2/grid2.md

---

# Grid version 2

<p class="description">The responsive layout grid adapts to screen size and orientation, ensuring consistency across layouts.</p>

{{"component": "modules/components/ComponentLinkHeader.js", "design": false}}

The `Grid` component works well for a layout with known columns. The columns can be configured in multple breakpoints which you have to specify the column span of each child.

## How it works

The grid system is implemented with the `Grid` component:

- It uses [CSS's Flexible Box module](https://www.w3.org/TR/css-flexbox-1/) for high flexibility.
- The grid is always a flex item. Use the `container` prop to add flex container to it.
- Item widths are set in percentages, so they're always fluid and sized relative to their parent element.
- There are five default grid breakpoints: xs, sm, md, lg, and xl. If you need custom breakpoints, check out [custom breakpoints grid](#custom-breakpoints).
- Integer values can be given to each breakpoint, indicating how many of the 12 available columns are occupied by the component when the viewport width satisfies the [breakpoint constraints](/material-ui/customization/breakpoints/#default-breakpoints).

> - It uses negative margin and padding technique to create [gap-like](https://developer.mozilla.org/en-US/docs/Web/CSS/gap) between children.

Reviewer: ```suggestion
- It uses negative margins and padding to create gaps between children, which behave similarly to [the `gap` CSS property](https://developer.mozilla.org/en-US/docs/Web/CSS/gap).
```

> - It **does not** have the concept of rows. Meaning, you can't make the children span to multiple rows. If you need to do that, we recommend to use [CSS Grid](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout) instead.

Reviewer: ```suggestion
- It does _not_ support row spanning. Children elements cannot span multiple rows. We recommend using [CSS Grid](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout) if you need this functionality.
```

> - It **does not** offer auto-placement children feature. It will try to fit the children one by one and if there is not enough space, the rest of the children will start on the next line and so on. If you need the auto-placement feature, we recommend to use [CSS Grid](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Auto-placement_in_CSS_Grid_Layout) instead.

Reviewer: ```suggestion
- It does _not_ automatically place children. It will try to fit the children one by one, and if there is not enough space, the rest of the children will start on the next line, and so on. If you need auto-placement, we recommend using [CSS Grid](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Auto-placement_in_CSS_Grid_Layout) instead.
```

:::warning

> ⚠️ The `Grid` component shouldn't be confused with a data grid; it is closer to a layout grid. For a data grid head to [the `DataGrid` component](/x/react-data-grid/).

Reviewer: ```suggestion
The `Grid` component is a _layout_ grid, not a _data_ grid. 
If you need a data grid, check out [the MUI X `DataGrid` component](/x/react-data-grid/).
```

:::

## What's changed

> We build Grid v2 from scratch in order to:

Reviewer 2: I would add note at the top of the page about why this components exists and is named Grid v2 :)

> We build Grid v2 from scratch in order to:

Reviewer: ```suggestion
We built `Grid` v2 from scratch in order to:
```

- Fix [known issues](https://github.com/mui/material-ui/pull/32746) introduced in Material UI v5.

> - Simplify the logic with CSS variables which removes unnecessary `item` prop and reduce CSS specificity.

Reviewer: ```suggestion
- Simplify the logic with CSS variables, removing the unnecessary `item` prop and reducing CSS specificity.
```

- Introduce a proper fix for [preventing a scrollbar](#prevent-scrollbar) by switching between negative margin approaches.

> - Equally set negative margin on all sides of the grid container by default.

Reviewer: ```suggestion
- Set negative margins of equal size on all sides of the grid container by default.
```

> Changing Grid from v1 to v2 is considered a breaking change, please follow a [migration guide](/material-ui/migration/migration-grid-v2/) for more details.

Reviewer: ```suggestion
Updating the `Grid` from v1 to v2 is a breaking change. 
Visit the [Grid v2 migration guide](/material-ui/migration/migration-grid-v2/) for more details.
```

:::info

> In the next major release of Material UI, we will make the Grid v2 stable and deprecate the v1.

Reviewer: ```suggestion
In the next major release of Material UI, Grid v2 will be stable and v1 will be deprecated.
```

:::

## Fluid grids

Fluid grids use columns that scale and resize content. A fluid grid's layout can use breakpoints to determine if the layout needs to change dramatically.

### Basic grid

> In order to create a grid layout, you need a container. Use `container` prop to create a grid container that wraps the grid items (the `Grid` is always an item).

Reviewer: ```suggestion
In order to create a grid layout, you need a container. 
Use the `container` prop to create a grid container that wraps the grid items (the `Grid` is always an item).
```

> Column widths are integer values between 1 and 12; they apply at any breakpoint and indicate how many columns are occupied by the component.

Reviewer: ```suggestion
Column widths are integer values between 1 and 12. 
They can be applied at any breakpoint to indicate how many columns are occupied by the component.
```

## Offset

Move the item to the right by using offset props which can be:

- number, for example, `mdOffset={2}` - when used the item is moved to the right by 2 columns starts from `md` breakpoint and up.
- `"auto"` - when used, the item is moved to the right edge of the grid container.

{{"demo": "OffsetGrid.js", "bg": true}}

## Custom breakpoints

If you specify custom breakpoints to the theme, you can use those names as grid item props in responsive values.

```js
import { ThemeProvider, createTheme } from '@mui/material/styles';

function Demo() {
  return (
    <ThemeProvider
      theme={createTheme({
        breakpoints: {
          values: {
            laptop: 1024,
            tablet: 640,
            mobile: 0,
            desktop: 1280,
          },
        },
      })}
    >
      <Grid container spacing={{ mobile: 1, tablet: 2, laptop: 3 }}>
        {Array.from(Array(4)).map((_, index) => (
          <Grid mobile={6} tablet={4} laptop={3} key={index}>
            <div>{index + 1}</div>
          </Grid>
        ))}
      </Grid>
    </ThemeProvider>
  );
}
```

:::info
The custom breakpoints affect both the size and offset props:

```diff

> - <Grid xs={6} xsOffset={2} sm={4} smOffset={2} md={3} mdOffset={3}>
> + <Grid mobile={6} mobileOffset={2} tablet={4} tabletOffset={2} laptop={3} laptopOffset={3}>

Author: ```suggestion
- <Grid xs={6} xsOffset={2}>
+ <Grid mobile={6} mobileOffset={2}>
```

Author: follows https://github.com/mui/material-ui/pull/33527

{{"demo": "ResponsiveGrid.js", "bg": true}}

Responsive values is supported by:

- `columns`
- `columnSpacing`
- `direction`
- `rowSpacing`
- `spacing`
- all the [other props](#system-props) of the system

## Auto-layout

The Auto-layout makes the _items_ equitably share the available space.
That also means you can set the width of one _item_ and the others will automatically resize around it.

{{"demo": "AutoGrid.js", "bg": true}}

### Variable width content

Set one of the size breakpoint props to `"auto"` instead of `true` / a `number` to size
a column based on the natural width of its content.

{{"demo": "VariableWidthGrid.js", "bg": true}}

## Nested Grid

The grid container that renders inside another grid container is a nested grid which inherits the [`columns`](#columns) and [`spacing`](#spacing) from the top. The deep nested grid will inherit the props from the upper nested grid if it receives those props.

{{"demo": "NestedGrid.js", "bg": true}}

## Columns

You can change the default number of columns (12) with the `columns` prop.

{{"demo": "ColumnsGrid.js", "bg": true}}

## Offset

> Move the item to the right by using offset props which can be:

Reviewer: ```suggestion
Offset props (such as `smOffset`, `mdOffset`) push an item to the right side of the grid.
These props accept:
```

> - number, for example, `mdOffset={2}` - when used the item is moved to the right by 2 columns starts from `md` breakpoint and up.

Reviewer: ```suggestion
- numbers—for example, `mdOffset={2}` pushes an item two columns to the right when the viewport size is equal to or greater than the `md` breakpoint.
```

> - `"auto"` - when used, the item is moved to the right edge of the grid container.

Reviewer: ```suggestion
- `"auto"`—this pushes the item to the far right side of the grid container.
```

> 

Reviewer: ```suggestion

The demo below illustrates how to use the offset props:

```

{{"demo": "OffsetGrid.js", "bg": true}}

## Custom breakpoints

> If you specify custom breakpoints to the theme, you can use those names as grid item props in responsive values.

Reviewer: ```suggestion
If you specify custom breakpoints in the theme, you can use those names as grid item props in responsive values:
```

```js
import { ThemeProvider, createTheme } from '@mui/material/styles';

function Demo() {
  return (
    <ThemeProvider
      theme={createTheme({
        breakpoints: {
          values: {
            laptop: 1024,
            tablet: 640,
            mobile: 0,
            desktop: 1280,
          },
        },
      })}
    >
      <Grid container spacing={{ mobile: 1, tablet: 2, laptop: 3 }}>
        {Array.from(Array(4)).map((_, index) => (
          <Grid mobile={6} tablet={4} laptop={3} key={index}>
            <div>{index + 1}</div>
          </Grid>
        ))}
      </Grid>
    </ThemeProvider>
  );
}
```

:::info

> The custom breakpoints affect both the size and offset props:

Reviewer: ```suggestion
Custom breakpoints affect both size and offset props:
```

```diff
- <Grid xs={6} xsOffset={2}>
+ <Grid mobile={6} mobileOffset={2}>
```

:::

### TypeScript

> You have to set module augmentation on the theme breakpoints interface. The properties with `true` value will appear as `{key}`(size prop) and `{key}Offset`(offset prop).

Reviewer: ```suggestion
You have to set module augmentation on the theme breakpoints interface. 
Properties set to `true` will appear as `{key}`(size prop) and `{key}Offset`(offset prop).
```

```ts
declare module '@mui/system' {
  interface BreakpointOverrides {
    // Your custom breakpoints
    laptop: true;
    tablet: true;
    mobile: true;
    desktop: true;
    // Remove default breakpoints
    xs: false;
    sm: false;
    md: false;
    lg: false;
    xl: false;
  }
}
```

> ## Prevent scrollbar

Reviewer: ```suggestion
## Disable the scrollbar
```

If you use grid as a container in a small viewport, you might see a horizontal scrollbar because the negative margin is applied on all sides of the grid container.

> To prevent the scrollbar, set `disableEqualOverflow` prop to `true`. It will enable negative margin only on the top and left sides of the grid which remove overflow on the right-hand side.

Reviewer: ```suggestion
To disable this scrollbar, set the `disableEqualOverflow` prop to `true`. 
This removes the negative margins from the bottom and right sides of the grid to prevent overflow.

The demo below shows how this works:
```

{{"demo": "OverflowGrid.js", "bg": true}}

:::warning

> You should avoid adding borders or background to the grid when `disableEqualOverflow: true` because the negative margin (applied only at the top and left sides) makes the grid visually misaligned.

Reviewer: ```suggestion
You should avoid adding borders and backgrounds to the grid when `disableEqualOverflow` is `true` because the negative margin (applied only at the top and left sides) causes the grid to be visually misaligned.
```

:::

## Customization

### Centered elements

> If you want to make the content of the grid item centered, specify `display="flex"` directly to the grid item. Then, use `justifyContent` and/or `alignItems` to adjust the position of the content.

Reviewer: ```suggestion
To center a grid item's content, specify `display="flex"` directly on the item. 
Then use `justifyContent` and/or `alignItems` to adjust the position of the content, as shown below:
```

{{"demo": "CenteredElementGrid.js", "bg": true}}

:::warning

> Using `container` prop would not work because the grid container is designed to wrap only the grid items, not random elements.

Reviewer: ```suggestion
Using the `container` prop does not work in this situation because the grid container is designed exclusively to wrap grid items.
It cannot wrap other elements.
```

:::

> ### Full bordered

Reviewer: ```suggestion
### Full border
```

{{"demo": "FullBorderedGrid.js"}}

> ### Half bordered

Reviewer: ```suggestion
### Half border
```

{{"demo": "HalfBorderedGrid.js"}}

## Limitations

> ### direction column and column-reverse

Reviewer: ```suggestion
### Column direction and reversing
```

> The column width (`xs`, ..., `xl`) and offset props are **not supported** within `direction="column"` and `direction="column-reverse"` containers.

Reviewer: ```suggestion
The column width (`xs`, ..., `xl`) and offset props are _not_ supported within containers that use `direction="column"` or `direction="column-reverse"`.
```

> They define the number of grids the component will use for a given breakpoint. They are intended to control **width** using `flex-basis` in `row` containers but they will impact height in `column` containers.

Reviewer: "They" is ambiguous here. Does it refer to the size & offset props, or the direction=x properties?

Author: it refers to the size & offset props.

Reviewer: How about this? 

```suggestion
Size and offset props define the number of columns the component will use for a given breakpoint. 
They are intended to control the width using `flex-basis` in `row` containers, but they will impact the height in `column` containers.
```
